templates/products/products-listing-row.php

```diff
@@ -221,34 +225,32 @@
         } elseif ( dokan_get_prop( $product, 'product_type', 'get_type' ) === 'external' ) {
-            echo '<span class="product-type tips external" title="' . esc_html__( 'External/Affiliate', 'dokan-lite' ) . '"></span>';
+            $product_type = '<span class="product-type tips external" title="' . esc_html__( 'External/Affiliate', 'dokan-lite' ) . '"></span>';
         } elseif ( dokan_get_prop( $product, 'product_type', 'get_type' ) === 'simple' ) {
             if ( $product->is_virtual() ) {
-                echo '<span class="product-type tips virtual" title="' . esc_html__( 'Virtual', 'dokan-lite' ) . '"></span>';
+                $product_type = '<span class="product-type tips virtual" title="' . esc_html__( 'Virtual', 'dokan-lite' ) . '"></span>';
             } elseif ( $product->is_downloadable() ) {
-                echo '<span class="product-type tips downloadable" title="' . esc_html__( 'Downloadable', 'dokan-lite' ) . '"></span>';
+                $product_type = '<span class="product-type tips downloadable" title="' . esc_html__( 'Downloadable', 'dokan-lite' ) . '"></span>';
             } else {
-                echo '<span class="product-type tips simple" title="' . esc_html__( 'Simple', 'dokan-lite' ) . '"></span>';
+                $product_type = '<span class="product-type tips simple" title="' . esc_html__( 'Simple', 'dokan-lite' ) . '"></span>';
             }
         } elseif ( dokan_get_prop( $product, 'product_type', 'get_type' ) === 'variable' ) {
-            echo '<span class="product-type tips variable" title="' . esc_html__( 'Variable', 'dokan-lite' ) . '"></span>';
+            $product_type = '<span class="product-type tips variable" title="' . esc_html__( 'Variable', 'dokan-lite' ) . '"></span>';
         } else {
             // Assuming that we have other types in future
-            echo '<span class="product-type tips ' . esc_attr( dokan_get_prop( $product, 'product_type', 'get_type' ) ) . '" title="' . esc_attr( ucfirst( dokan_get_prop( $product, 'product_type', 'get_type' ) ) ) . '"></span>';
+            $product_type = '<span class="product-type tips ' . esc_attr( dokan_get_prop( $product, 'product_type', 'get_type' ) ) . '" title="' . esc_attr( ucfirst( dokan_get_prop( $product, 'product_type', 'get_type' ) ) ) . '"></span>';
         }
 
         /**
-         * Fire an action to add extra content after product type column in product listing table
+         * Render product listing product type symbol.
          *
          * @since 3.11.5
          *
-         * @param \WC_Product $product Current product.
-         *
-         * @return void
+         * @param string $product_type
          */
-        do_action( 'dokan_product_list_table_after_column_content_type', $product );
+        echo apply_filters( 'dokan_product_listing_product_type', $product_type, $product );
```

Comment:
We should not remove an action without deprecating it first.

Please rethink it and also preserve the doc comments appropriately with all the Params.

```php
echo apply_filters( 'dokan_product_listing_product_type', $product_type, $product );
do_action( 'dokan_product_list_table_after_column_content_type', $product );

```